### PR TITLE
 [ZEPPELIN-4187] Bump up version of Scala from 2.11.8 to 2.11.12

### PR DESCRIPTION
**Pull request by prabhjyotsingh**
### What is this PR for?
This is to bump up the version of Scala from 2.11.8 to 2.11.12. 
This is a minor version upgrade to the latest fixes from Scala.

### What type of PR is it?
[Improvement]

### What is the Jira issue?
*  [ZEPPELIN-4187](https://jira.apache.org/jira/browse/ZEPPELIN-4187)

### How should this be tested?
* CI should be green

### Screenshots (if appropriate)
N/A

### Questions:
* Does the licenses files need update? n/a
* Is there breaking changes for older versions? n/a
* Does this needs documentation? n/a

### CODE REVIEW

**Comment by prabhjyotsingh** [comment]
@felixcheung @zjffdu  can you help review this.
Also, should we upgrade to the latest 2.11.12?

**Comment by zjffdu** [comment]
I believe so, that latest spark support scala 2.11.12 https://github.com/apache/spark/blob/branch-2.4/pom.xml#L158

**Comment by prabhjyotsingh** [comment]
Thank you @zjffdu  that definitely makes sense, let me update the version to `2.11.12` and change the PR title as well.

File: pom.xml [hunk 1]
```diff
@@ -750,7 +750,7 @@
     <profile>
       <id>scala-2.11</id>
       <properties>
-        <scala.version>2.11.8</scala.version>
+        <scala.version>2.11.12</scala.version>
```
Comment (zjffdu) [inline pom.xml]: I think we just only need to change scala version in spark interpreter. Changing it in project root's `pom.xml` will affect all the interpreters, this may introduce other issues.

**Comment by prabhjyotsingh** [comment]
We need to upgrade Scala version for Flink as well, or should it run with an older version of Scala (2.11.8)?

File: pom.xml (hunk 1 above)
Comment (prabhjyotsingh) [reply pom.xml]: I see all of Casandra, Flink, and Ignite will get the version upgrade because of this.
Sure, let me revert this change.

**Comment by felixcheung** [comment]
flink can run with 2.11.12 https://github.com/apache/flink/blob/master/pom.xml#L119
but maybe we need to update the flink version - we have 1.7

File: flink/pom.xml [hunk 2]
```diff
@@ -36,11 +36,12 @@
...
     <interpreter.name>flink</interpreter.name>
-    <flink.version>1.7.1</flink.version>
+    <flink.version>1.7.2</flink.version>
     <flink.akka.version>2.3.7</flink.akka.version>
-    <scala.macros.version>2.0.1</scala.macros.version>
+    <scala.macros.version>2.1.1</scala.macros.version>
     <scala.binary.version>2.11</scala.binary.version>
-    <scala.version>2.11.8</scala.version>
+    <scala.version>2.11.12</scala.version>
+    <spark.version>2.4.0</spark.version>
```
Comment (felixcheung) [inline flink/pom.xml]: why is this here?

File: flink/src/main/scala/org/apache/zeppelin/flink/FlinkScalaInterpreter.scala [hunk 3]
```diff
@@ -30,18 +29,20 @@ import org.apache.flink.runtime.minicluster.MiniCluster
 import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
 import org.apache.flink.table.api.TableEnvironment
 import org.apache.flink.table.api.scala.{BatchTableEnvironment, StreamTableEnvironment}
+import org.apache.spark.repl.SparkILoop
```
Comment (felixcheung) [inline flink/src/main/scala/org/apache/zeppelin/flink/FlinkScalaInterpreter.scala]: ???

File: flink/src/main/scala/org/apache/zeppelin/flink/FlinkScalaInterpreter.scala [hunk 4]
```diff
@@ -231,3 +233,71 @@ class FlinkScalaInterpreter(val properties: Properties) {
...
+    * https://github.com/scala/scala/blob/v2.11.8/src/repl/scala/tools/nsc/interpreter/ILoop.scala
+    * https://github.com/scala/scala/blob/v2.11.12/src/repl/scala/tools/nsc/interpreter/ILoop.scala
+    *
+    * See also ZEPPELIN-3810/ZEPPELIN-4187.
+    */
+  private def loopPostInit(interpreter: FlinkScalaInterpreter): Unit = {
+    import StdReplTags._
+    import scala.reflect.{classTag, io}
+
+    val sparkILoop = interpreter.flinkILoop
```
Comment (felixcheung) [inline flink/src/main/scala/org/apache/zeppelin/flink/FlinkScalaInterpreter.scala]: `sparkILoop`?

Comment (prabhjyotsingh) [reply flink/src/main/scala/org/apache/zeppelin/flink/FlinkScalaInterpreter.scala]: `callMethod(flinkILoop, "scala$tools$nsc$interpreter$ILoop$$loopPostInit")` was failing so got it replaced with the help of ZEPPELIN-3810, let me fix the variable name.

**Comment by prabhjyotsingh** [comment]
@felixcheung yes I've upgraded Flink from 1.7.1 to 1.7.2 as well. To upgrade it to 1.8.0 will be slightly tricky, I can do it on a follow-up PR. Or does it makes more sense to revert all of the Flink changes from here and implement it on a different PR?

**Comment by prabhjyotsingh** [comment]
CI is failing for unrelated issue. Is this good to be merged?

**Comment by felixcheung** [comment]
it's definitely strange that the flink interpreter is depending on spark
eg. `org.apache.spark.repl`

**Comment by zjffdu** [comment]
As I mentioned before, this PR's scope should be only limited to spark module. Other module should not be affected.

**Comment by prabhjyotsingh** [comment]
Sure, I've just reverted all of Flink changes as well.

**Comment by prabhjyotsingh** [comment]
@zjffdu these two changes looks ok, right?

**Comment by prabhjyotsingh** [comment]
Thank you @zjffdu 

Will merge this, if no more discussion.
